### PR TITLE
feat: expand fsm instrumentation

### PR DESCRIPTION
**Pull request by jonathanj-square**
- extends the `fsm_instances` table scheme to help distinguish between insert and update.
- track active fsm transitions
- track fsm transition attempts

Issue: https://github.com/TBD54566975/ftl/issues/2194

### CODE REVIEW

File: backend/controller/dal/fsm_test.go [hunk 1]
```diff
@@ -15,6 +16,8 @@ import (
 )
 
 func TestSendFSMEvent(t *testing.T) {
+	_ = observability.InitControllerObservability()
```
Comment (alecthomas) [inline backend/controller/dal/fsm_test.go:19]: I wasn't sure it was a good idea, but I think that having to do this everywhere is a good argument for statically initialising observability at `init()` time.

Comment (wesbillman) [reply backend/controller/dal/fsm_test.go:19]: Yeah totally agree. That would not be fun to have to remember this for every test that might have a metric. `init()` sounds good to me. Will probably clean up quite a bit of this plumbing too.

File: backend/controller/observability/observability.go [hunk 2]
```diff
@@ -2,10 +2,22 @@ package observability
...
+	if err != nil {
 		return fmt.Errorf("could not initialize controller metrics: %w", err)
 	}
 
 	return nil
 }
+
+func FSM() *FSMMetrics {
+	return fsm
+}
```
Comment (wesbillman) [inline backend/controller/observability/observability.go:23]: We can probably just expose the public var here vs. having a wrapper func. This is likely more true with an `init()` approach.

Comment (jonathanj-square) [reply backend/controller/observability/observability.go]: feedback adopted

File: backend/controller/observability/fsm.go [hunk 3]
```diff
@@ -2,53 +2,91 @@ package observability
...
 
-var fsmCounters = struct {
-	instancesActive metric.Int64UpDownCounter
-}{}
+func initFSMMetrics() (*FSMMetrics, error) {
+	result := &FSMMetrics{}
 
-func InitFSMMetrics() error {
+	var errs error
 	var err error
```
Comment (deniseli) [inline backend/controller/observability/fsm.go:34]: Would you mind removing this `var err`? Generally, you only initialize vars in Go when you have to (e.g. to get an address of a particular type), so for this, we can just use `:=` for the first `result.blah, err := `

File: backend/controller/observability/fsm.go [hunk 4]
```diff
@@ -2,53 +2,91 @@ package observability
...
-		metric.WithDescription("counts the number of active FSM instances"))
+	result.meter = otel.Meter("ftl.fsm")
 
-	if err != nil {
-		return fmt.Errorf("could not initialize fsm metrics: %w", err)
+	counter := fmt.Sprintf("%s.instances.active", fsmMeterName)
+	if result.instancesActive, err = result.meter.Int64UpDownCounter(
+		counter,
+		metric.WithDescription("counts the number of active FSM instances")); err != nil {
+		errs = errors.Join(errs, fmt.Errorf("%q counter init failed; falling back to noop: %w", counter, err))
```
Comment (deniseli) [inline backend/controller/observability/fsm.go:42]: Just to dedup the repeated code, what do you think about something like:
```
    errs = joinInitErrors(errs, err, counter)
    ...
}

func joinInitErrors(...) error {
    return errors.Join(errs, fmt.Errorf("%q counter init failed; falling back to noop: %w", counter, err))
}
```

**Review by deniseli — APPROVED** [review]

File: backend/controller/observability/fsm.go [hunk 5]
```diff
@@ -2,53 +2,91 @@ package observability
...
+	result := &FSMMetrics{}
 
-func InitFSMMetrics() error {
+	var errs error
 	var err error
 
-	fsmCounters.instancesActive, err = fsmMeter.Int64UpDownCounter(
-		fmt.Sprintf("%s.instances.active", fsmMeterName),
-		metric.WithDescription("counts the number of active FSM instances"))
+	result.meter = otel.Meter("ftl.fsm")
```
Comment (deniseli) [inline backend/controller/observability/fsm.go:36]: `result.meter = otel.Meter(fsmMeterName)`

File: backend/controller/observability/fsm.go (hunk 3 above)
Comment (deniseli) [reply backend/controller/observability/fsm.go:34]: Ah nvm, I see where you need it now

File: backend/controller/observability/fsm.go (hunk 5 above)
Comment (jonathanj-square) [reply backend/controller/observability/fsm.go]: adopted

File: backend/controller/observability/fsm.go (hunk 4 above)
Comment (jonathanj-square) [reply backend/controller/observability/fsm.go]: adopted

**Pull request merged**